src/System.Private.CoreLib/src/System/Runtime/InteropServices/GCHandle.CoreRT.cs

```diff
@@ -0,0 +1,71 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Internal.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    public partial struct GCHandle
+    {
+        private IntPtr InternalAlloc(object value, GCHandleType type)
+        {
+            if (type == GCHandleType.Pinned)
+            {
+                GCHandleValidatePinnedObject(value);
+            }
+
+            return RuntimeImports.RhHandleAlloc(value, type);
+        }
+
+        private void InternalFree(IntPtr handle) => RuntimeImports.RhHandleFree(handle);
+
+        private object InternalGet(IntPtr handle) => RuntimeImports.RhHandleGet(handle);
+
+        private void InternalSet(IntPtr handle, object value, bool isPinned)
+        {
+            if (isPinned)
+            {
+                GCHandleValidatePinnedObject(value);
+            }
+
+            RuntimeImports.RhHandleSet(handle, value);
+        }
+
+        private unsafe IntPtr InternalAddrOfPinnedObject(IntPtr handle)
+        {
+            // Get the address of the pinned object.
+            // The layout of String and Array is different from Object
+
+            object target = InternalGet(handle);
```

Comment:
I think all that's used here is available in shared.

Reply:
I commented on that in the coreclr PR... I tried using it in coreclr but it made AddrOfPinnedObject 2x slower on my machine for non-string objects.

Reply:
I guess that it is the array cast that's expensive. You can try to use the same trick as what is used for `Memory<T>` for a cheap array check, something like:

```
if (RuntimeHelpers.ObjectHasComponentSize(target))
{
    if (target.GetType() == typeof(string))
        return Unsafe.As<string>(target).GetRawStringData();

    Debug.Assert(target is Array);
    return Unsafe.As<Array>(target).GetRawArrayData();
}
```

Reply:
That helps, but doesn't complete eliminate the regression.  With this microbenchmark:
```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System;
using System.Runtime.InteropServices;

[InProcess]
public class Test
{
    public static void Main() => BenchmarkRunner.Run<Test>();

    private static GCHandle s_pinnedString = GCHandle.Alloc(new string('c', 42), GCHandleType.Pinned);
    private static GCHandle s_pinnedInt32Box = GCHandle.Alloc(42, GCHandleType.Pinned);
    private static GCHandle s_pinnedArray = GCHandle.Alloc(new byte[42], GCHandleType.Pinned);

    [Benchmark] public IntPtr AddrOfPinnedString() => s_pinnedString.AddrOfPinnedObject();
    [Benchmark] public IntPtr AddrOfPinnedInt32Box() => s_pinnedInt32Box.AddrOfPinnedObject();
    [Benchmark] public IntPtr AddrOfPinnedArray() => s_pinnedArray.AddrOfPinnedObject();
}
```
With the existing InternalCall, I get results like:
```
               Method |     Mean |     Error |    StdDev |
--------------------- |---------:|----------:|----------:|
   AddrOfPinnedString | 3.455 ns | 0.0321 ns | 0.0285 ns |
 AddrOfPinnedInt32Box | 3.034 ns | 0.0290 ns | 0.0242 ns |
    AddrOfPinnedArray | 3.767 ns | 0.0313 ns | 0.0292 ns |
```
and with this implementation:
```C#
public IntPtr AddrOfPinnedObject()
{
    // Check if the handle was not a pinned handle.
    // You can only get the address of pinned handles.
    ThrowIfInvalid();
    if (!IsPinned)
    {
        ThrowHelper.ThrowInvalidOperationException_HandleIsNotPinned();
    }

    // Get the address.

    object target = InternalGet(GetHandleValue());
    if (target is null)
    {
        return default;
    }

    unsafe
    {
        if (RuntimeHelpers.ObjectHasComponentSize(target))
        {
            if (target.GetType() == typeof(string))
            {
                return (IntPtr)Unsafe.AsPointer(ref Unsafe.As<string>(target).GetRawStringData());
            }

            Debug.Assert(target is Array);
            return (IntPtr)Unsafe.AsPointer(ref Unsafe.As<Array>(target).GetRawArrayData());
        }

        return (IntPtr)Unsafe.AsPointer(ref target.GetRawData());
    }
}
```
I get results like:
```
               Method |     Mean |     Error |    StdDev |
--------------------- |---------:|----------:|----------:|
   AddrOfPinnedString | 2.459 ns | 0.0762 ns | 0.0713 ns |
 AddrOfPinnedInt32Box | 3.010 ns | 0.0557 ns | 0.0521 ns |
    AddrOfPinnedArray | 4.437 ns | 0.0742 ns | 0.0658 ns |
```
so strings get faster, boxes stay the same, and arrays get slower.  Anything we could do about the array case?  Is there something about GetRawArrayData that's causing an issue and could be improved?  Or did I just do something silly?